### ISSUE TITLE
Make thrift transport pool more concurrent

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#1122]
https://issues.apache.org/jira/browse/ACCUMULO-4821

**Comment by hkeebler** [comment]
Is this a straight forward as setting the below and removing the synchronized blocks? 
 ```
   Queue<CachedConnection> unreserved = new ConcurrentLinkedQueue<>();
    Map<CachedTTransport,CachedConnection> reserved = new ConcurrentHashMap<>();
.................. 
  private Map<ThriftTransportKey,CachedConnections> cache = new ConcurrentHashMap<>();
```
I do not see any unit tests for this.  There is a TransportCachingIT but I'll need to review to see if it adequately test this.

**Comment by ctubbsii** [comment]
TransportCachingIT used to hang indefinitely before we added a timeout to it to force it to be killed if it took too long. It's quite possible there's a concurrency bug in there somewhere, but I'm not familiar with that code.

**Comment by keith-turner** [comment]
> Is this a straight forward as setting the below and removing the synchronized blocks?

I looked at the code and I think more may be required.  There is a thread that closes idle unreserved connections.  Would need to ensure there are no race conditions with this code and reserving a connection.  That is one example where removing the sync may introduce race conditions.

**Comment by hkeebler** [comment]
Thanks.   I will carefully review the synchronized blocks of code for refactoring.

**Comment by lbschanno** [comment]
@keith-turner is this fixed by PR #1241?

**Comment by keith-turner** [comment]
> @keith-turner is this fixed by PR #1241?

No, there were some nice improvements there.  However its still globally synchronized. Those changes sped up some of the work done while the lock is held.

**Comment by lbschanno** [comment]
@keith-turner understood, thanks for explaining.

**Comment by hkeebler** [comment]
Go for it @lbschanno !

**Comment by lbschanno** [comment]
@keith-turner I've been playing around with a couple of different approaches to this. One possible approach is using a separate ConcurrentHashMap of ThriftTransportKeys to ReentrantLocks to support key-pair granularity when fetching and modifying the connections for a particular tserver. The following example shows how the synchronized block is removed from `ThriftTransportKeys.getTransport()`. 

```java
public class ThriftTransportPool {
  
  private ConcurrentHashMap<ThriftTransportKey, CachedConnections> cache = new ConcurrentHashMap<>();
  private ConcurrentHashMap<ThriftTransportKey, ReentrantLock> locks = new ConcurrentHashMap<>();
  
  private ReentrantLock getLock(final ThriftTransportKey key) {
    return locks.computeIfAbsent(key, k -> new ReentrantLock());
  }
  
  private TTransport getTransport(ThriftTransportKey cacheKey) throws TTransportException {
    // compute hash code outside of lock, this lowers the time the lock is held
    cacheKey.precomputeHashCode();
    
    Lock lock = getLock(cacheKey);
    lock.lock();
    
    try {
      // atomically reserve location if it exist in cache
      CachedConnection cachedConnection =
          getCache().computeIfAbsent(cacheKey, ck -> new CachedConnections()).reserveAny();
      if (cachedConnection != null) {
        log.trace("Using existing connection to {}", cacheKey.getServer());
        return cachedConnection.transport;
      }
    } finally {
      lock.unlock();
    }

    return createNewTransport(cacheKey);
  }
}
```

But before I go too far down this path, I wanted to check if you see any issues with this approach.

**Comment by keith-turner** [comment]
@lbschanno that sounds reasonable.  One thing I am curious about is how you plan to remove the lock safely when the cached connections set goes empty.  This needs to be done in such a way that threads never end up using different locks.

**Comment by lbschanno** [comment]
@keith-turner I've submitted draft PR #1374 for you to take a look at when you have a chance.

**Comment by keith-turner** [comment]
Closed by 2e5182b
